### ISSUE TITLE
Dispose never called during partial enumeration

### ISSUE THREAD

**Issue by TomPeters** [tyrotoxin/AsyncEnumerable#4]
Consider the following:

``` C#
void Main()
{
	GetEnumerable().TakeAsync(2).ForEachAsync(i => i.Dump());
}

public IAsyncEnumerable<int> GetEnumerable()
{
	return new AsyncEnumerable<int>(async yield =>
	{
		using (new MyDisposable())
		{
			await yield.ReturnAsync(1);
			await yield.ReturnAsync(2);
			await yield.ReturnAsync(3);
		}
	});
}

public class MyDisposable : IDisposable
{
	public MyDisposable()
	{
		"Created".Dump();
	}
	
	public void Dispose()
	{
		"Disposed".Dump();
	}
}
```

The output of this is:
```
Created
1
2
```

Compare this to the non-async equivalent:
``` C#
void Main()
{
	foreach(var i in GetEnumerable().Take(2))
		i.Dump();
}

public IEnumerable<int> GetEnumerable()
{
	using (new MyDisposable())
	{
		yield return 1;
		yield return 2;
		yield return 3;
	}
}

public class MyDisposable : IDisposable
{
	public MyDisposable()
	{
		"Created".Dump();
	}

	public void Dispose()
	{
		"Disposed".Dump();
	}
}
```

In this case, the output is 
```
Created
1
2
Disposed
```

So in the async version, dispose is not guaranteed to be called, while it is in the sync version.

**Comment by tyrotoxin** [comment]
Thanks @TomPeters, I'll take a look at that.

**Comment by tyrotoxin** [comment]
@TomPeters, thanks for contributing!
Fix is available in v1.2.3 and is already on nuget.org.

**Comment by TomPeters** [comment]
No problem, thanks for the quick response 👍
